From 9f4c1f1305b4a9c66df0f14a8fb251efdd12c2c6 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:12:48 -0400
Subject: [powerpc] fadump: Cleaning up inconsistent NULL checks

Message-id: <1410545655-205645-540-git-send-email-dzickus@redhat.com>
Patchwork-id: 94571
O-Subject: [RHEL7 PATCH 539/626] arch: powerpc/fadump: Cleaning up inconsistent NULL checks
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit b717d98543d90c50f398ddfe9665483d09adca7a
Author: Rickard Strandqvist <rickard_strandqvist@spectrumdigital.se>
Date:   Fri May 23 00:03:16 2014 +0200

    arch: powerpc/fadump: Cleaning up inconsistent NULL checks

    Cleaning up inconsistent NULL checks.
    There is otherwise a risk of a possible null pointer dereference.

    Was largely found by using a static code analysis program called cppcheck.

    Signed-off-by: Rickard Strandqvist <rickard_strandqvist@spectrumdigital.se>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/fadump.c b/arch/powerpc/kernel/fadump.c
index 0266774..d55e898 100644
--- a/arch/powerpc/kernel/fadump.c
+++ b/arch/powerpc/kernel/fadump.c
@@ -646,7 +646,7 @@ static int __init fadump_build_cpu_notes(const struct fadump_mem_struct *fdm)
   }
   /* Lower 4 bytes of reg_value contains logical cpu id */
   cpu = reg_entry->reg_value & FADUMP_CPU_ID_MASK;
-  if (!cpumask_test_cpu(cpu, &fdh->cpu_online_mask)) {
+  if (fdh && !cpumask_test_cpu(cpu, &fdh->cpu_online_mask)) {
    SKIP_TO_NEXT_CPU(reg_entry);
    continue;
   }
@@ -663,9 +663,11 @@ static int __init fadump_build_cpu_notes(const struct fadump_mem_struct *fdm)
  }
  fadump_final_note(note_buf);
 
- pr_debug("Updating elfcore header (%llx) with cpu notes\n",
+ if (fdh) {
+  pr_debug("Updating elfcore header (%llx) with cpu notes\n",
        fdh->elfcorehdr_addr);
- fadump_update_elfcore_header((char *)__va(fdh->elfcorehdr_addr));
+  fadump_update_elfcore_header((char *)__va(fdh->elfcorehdr_addr));
+ }
  return 0;
 
 error_out:
-- 
1.7.1